test/unit/TestVideoPipe.cpp

```diff
@@ -289,3 +289,280 @@ TEST(TestVideoPipe, pictureResizeScaleDownBy)
+        ASSERT_FALSE(pic);
+}
+
+TEST(TestVideoPipe, downratingCancelGrabInvalidFrame)
+{
+        int infps = 60;
+        int outfps = 30;
+        int clockrate = 90000;
+
+        int width = 640;
+        int height = 480;
+
+        VideoPipe vidPipe;
+        vidPipe.Init();
+        vidPipe.StartVideoCapture(width, height, outfps);
+
+        // Enqueue a few frames. First is consumed and second is dropped for downrating
+        for (int i = 0; i< 2; ++i )
+        {
+                VideoBuffer::shared sharedVidBuffer = std::make_shared<VideoBuffer>(width, height);
+                sharedVidBuffer->SetClockRate(clockrate);
+                sharedVidBuffer->SetTimestamp(i* (double)clockrate/infps);
+                auto queued = vidPipe.NextFrame(sharedVidBuffer);
+                ASSERT_EQ(queued, i+1);
+        }
+
+        {
+                auto pic = vidPipe.GrabFrame(0);
+                ASSERT_TRUE(pic);
+                ASSERT_EQ(pic->GetTimestamp(), 0U);
+        }
+
+        // Start grab frame that is waiting for the next input
+        // without cancel will block forever
+        auto blocking_grab_frame = std::async(std::launch::async, [&]()->auto{
+            return vidPipe.GrabFrame(0);
+        });
+
+        // Force the condition to be waiting on another frame OR a cancel.
+        // I.e. It would have woken up and handled second frame which it 
+        // decides to drop before we call cancel
+        using namespace std::chrono_literals;
+        std::this_thread::sleep_for(2s);
+        vidPipe.CancelGrabFrame();
+
+        auto pic = blocking_grab_frame.get();
+        ASSERT_FALSE(pic);
+}
+
+TEST(TestVideoPipe, dontBlockOnEmptyQueueWhenDownrating)
+{
+        int infps = 60;
+        int outfps = 30;
+        int clockrate = 90000;
+
+        int width = 640;
+        int height = 480;
+
+        VideoPipe vidPipe;
+        vidPipe.Init();
+        vidPipe.StartVideoCapture(width, height, outfps);
+
+        // Enqueue a few frames. First is consumed and second is dropped for downrating
+        for (int i = 0; i< 2; ++i )
+        {
+                VideoBuffer::shared sharedVidBuffer = std::make_shared<VideoBuffer>(width, height);
+                sharedVidBuffer->SetClockRate(clockrate);
+                sharedVidBuffer->SetTimestamp(i* (double)clockrate/infps);
+                auto queued = vidPipe.NextFrame(sharedVidBuffer);
+                ASSERT_EQ(queued, i+1);
+        }
+
+        {
+                auto pic = vidPipe.GrabFrame(0);
+                ASSERT_TRUE(pic);
+                ASSERT_EQ(pic->GetTimestamp(), 0U);
+        }
+
+        // Start grab frame that is waiting for the next input
+        // for at most 10msec, however with bug will wait forever
+        auto blocking_grab_frame = std::async(std::launch::async, [&]()->auto{
+            return vidPipe.GrabFrame(10);
+        });
+
+        auto pic = blocking_grab_frame.get();
+        ASSERT_FALSE(pic);
```

Comment:
I think it would be good to use
```suggestion
        ASSERT_EQ(nullptr, pic);
```

Reply:
done in all places and also inverse for ASSERT_TRUE(pic)